src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -70,22 +72,31 @@ public JsonClassInfo ElementClassInfo
 
         public void UpdateSortedPropertyCache(ref ReadStackFrame frame)
         {
-            // Check if we are trying to build the sorted cache.
-            if (frame.PropertyRefCache == null)
-            {
-                return;
-            }
+            Debug.Assert(frame.PropertyRefCache != null);
 
-            List<PropertyRef> newList;
-            if (_propertyRefsSorted != null)
-            {
-                newList = new List<PropertyRef>(_propertyRefsSorted);
-                newList.AddRange(frame.PropertyRefCache);
-                _propertyRefsSorted = newList.ToArray();
-            }
-            else
+            lock (_cacheLock)
             {
-                _propertyRefsSorted = frame.PropertyRefCache.ToArray();
+                if (_propertyRefsSorted != null)
+                {
+                    List<PropertyRef> replacementList = new List<PropertyRef>(_propertyRefsSorted);
+                    Debug.Assert(replacementList.Count <= PropertyNameCountCacheThreshold);
+
+                    // Verify replacementList will not become too large. This could occur in race conditions.
+                    List<PropertyRef> listToAppend = frame.PropertyRefCache;
+                    while (replacementList.Count + listToAppend.Count > PropertyNameCountCacheThreshold)
+                    {
+                        listToAppend.RemoveAt(listToAppend.Count - 1);
+                    }
```

Comment:
This is mutating frame.PropertyRefCache.  That's ok?  It won't be accessed concurrently by any other threads?

Reply:
What if replacementList.Count is already > PropertyNameCountCacheThreshold?  It's ok that we then end up clearing frame.PropertyRefCache?

Reply:
frame.PropertyRefCache is only accessed by a single thread -- the thread doing the deserialization. 

replacementList.Count will never be > PropertyNameCountCacheThreshold.

If replacementList.Count == PropertyNameCountCacheThreshold then yes frame.PropertyRefCache will be cleared. This cache is just a "level 1" cache meaning the real "cache"\metadata will return the correct property if not found in the level 1 cache.